core/templates/dev/head/services/site-analytics.service.ts

```diff
@@ -17,199 +17,224 @@
  * the learner and editor views.
  */
 
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+const constants = require( '../../../../../assets/constants');
+import { WindowRef } from 'services/contextual/window-ref.service';
+
+declare global {
+  interface Window {
+    ga: any;
+  }
+}
 // Service for sending events to Google Analytics.
 //
 // Note that events are only sent if the CAN_SEND_ANALYTICS_EVENTS flag is
 // turned on. This flag must be turned on explicitly by the application
 // owner in feconf.py.
 
-angular.module('oppia').factory('SiteAnalyticsService', [
-  '$window', 'CAN_SEND_ANALYTICS_EVENTS', function(
-      $window, CAN_SEND_ANALYTICS_EVENTS) {
-    // For definitions of the various arguments, please see:
-    // developers.google.com/analytics/devguides/collection/analyticsjs/events
-    var _sendEventToGoogleAnalytics = function(
-        eventCategory, eventAction, eventLabel) {
-      if ($window.ga && CAN_SEND_ANALYTICS_EVENTS) {
-        $window.ga('send', 'event', eventCategory, eventAction, eventLabel);
-      }
-    };
+@Injectable({
+  providedIn: 'root'
+})
+export class SiteAnalyticsService {
+  constructor(private windowRef: WindowRef) {}
+
+  // For definitions of the various arguments, please see:
+  // developers.google.com/analytics/devguides/collection/analyticsjs/events
+  _sendEventToGoogleAnalytics(
+      eventCategory: string, eventAction: string, eventLabel: string): void {
+    if (this.windowRef.nativeWindow.ga && constants.CAN_SEND_ANALYTICS_EVENTS) {
+      this.windowRef.nativeWindow.ga('send', 'event',
```

Comment:
Wow, this is working? I remember I tried it and it gave an error that nativeWindow does not have any property ga. I was stuck here.

Reply:
Yeah I have declared that explicilty above that Window does have a property ga. Line 26

Reply:
Umm, I'm sorry but I still didn't get how that's working by declaring window in global. Could you please elaborate for my knowledge?

Reply:
Yeah the only issue was since ga is not a predefined function, so TypeScript check would have failed. Since ga is being injected through a script most probably. So we just explicitly defined that hey, there's a thing called as ga in the window.